### PR TITLE
IGNITE-20926 Added 2.16 release notes.

### PR DESCRIPTION
**Pull request by NSAmelchev**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: RELEASE_NOTES.txt [hunk 1]
```diff
@@ -1,6 +1,138 @@
...
+* Added sensitive information hiding when IGNITE_TO_STRING_INCLUDE_SENSITIVE property is set to false.
+* Added support of setting partitions in SqlFieldsQuery.
+* Fixed CREATE TABLE ... WITH affinity_key option validation.
+* Fixed NPE on write query plan to performance statistics.
+* Fixed SQL metrics (added parsing metrics for the Calcite-based engine, fixed execution metrics).
+* Fixed failure on planning when cache size exceeds Integer.MAX_VALUE.
+* Fixed list of reserved keywords.
+* Fixed memory leak in MailboxRegistryImpl#remotes.
+* Fixed performance statistics gathering.
+* Fixed placeholder values from ModifyNode (backport of IGNITE-19096).
```
Comment (alex-plekhanov) [inline RELEASE_NOTES.txt:119]: Remove "backport of IGNITE-19096", it's a ticket from Ignite-3

File: RELEASE_NOTES.txt [hunk 2]
```diff
@@ -1,6 +1,138 @@
...
+* Removed CacheAtomicityMode#TRANSACTIONAL_SNAPSHOT cache mode.
+* Mixed atomicity cache groups now restricted, but IGNITE_ALLOW_MIXED_CACHE_GROUPS system option may temporarily allow them.
+* The ignite-ml, cassandra modules have been migrated to the Ignite extensions.
+
+Ignite:
+* Added ARM64 Docker images.
+* Added CLIENT_CONNECTION_ATTRIBUTES system view.
+* Added IndexQuery to performance statistics.
+* Added JMX commands invoker.
+* Added a description check if the command argument is enum type.
```
Comment (alex-plekhanov) [inline RELEASE_NOTES.txt:18]: It's hard to understand for end user what these lines are about.
Perhaps group all management API related tickets to somethink like "Added management API (common interface to ...)"?

File: RELEASE_NOTES.txt [hunk 3]
```diff
@@ -1,6 +1,138 @@
...
+* Added ability to configure metric exporter filter from Spring XML.
+* Added cache dumps.
+* Added client session outbound message queue size limit.
+* Added compact footer usage statistics to the idle_verify command.
+* Added descriptions and values to the properties list command.
+* Added documentation for the Index Reader Utility.
```
Comment (alex-plekhanov) [inline RELEASE_NOTES.txt:24]: Do we really need documentation tickets in release notes? Documentation even not in release package.

File: RELEASE_NOTES.txt [hunk 4]
```diff
@@ -1,6 +1,138 @@
...
+* Added extended debug logging of client connections.
+* Added limit support to IndexQuery.
+* Added logging of snapshot check procedure.
+* Added metrics for operations on secondary indexes.
+* Added node Consistent ID to the index validate command output.
+* Added number of partitions processed by the index worker to the output of the index commands.
+* Added performance statistics info about query properties and processed rows count by query.
+* Added support of SslContextFactory to CommandHandler (instead of GridSslBasicContextFactory).
+* Added support of cache group filter for dumps.
```
Comment (alex-plekhanov) [inline RELEASE_NOTES.txt:33]: We have "Added cache dumps" in this release, perhaps we don't need another tickets in release notes related to dumps

File: RELEASE_NOTES.txt [hunk 5]
```diff
@@ -1,6 +1,138 @@
 Apache Ignite Release Notes
 ===========================
 
+Apache Ignite In-Memory Distributed Database 2.16.0
+-----------------------------------------------------------
+
+(!) WARNINGS:
+* Operations on atomic caches within transactions are finally forbidden. The system property IGNITE_ALLOW_ATOMIC_OPS_IN_TX has been removed.
+* Removed CacheAtomicityMode#TRANSACTIONAL_SNAPSHOT cache mode.
+* Mixed atomicity cache groups now restricted, but IGNITE_ALLOW_MIXED_CACHE_GROUPS system option may temporarily allow them.
```
[the remark this answers is not in the document]
Comment (NSAmelchev) [reply RELEASE_NOTES.txt]: Done.

File: RELEASE_NOTES.txt [hunk 6]
```diff
@@ -1,6 +1,138 @@
...
+* The ignite-ml, cassandra modules have been migrated to the Ignite extensions.
+
+Ignite:
+* Added ARM64 Docker images.
+* Added CLIENT_CONNECTION_ATTRIBUTES system view.
+* Added IndexQuery to performance statistics.
```
[the remark this answers is not in the document]
Comment (NSAmelchev) [reply RELEASE_NOTES.txt]: Done.

File: RELEASE_NOTES.txt (hunk 3 above)
Comment (NSAmelchev) [reply RELEASE_NOTES.txt]: Done.

File: RELEASE_NOTES.txt (hunk 1 above)
Comment (NSAmelchev) [reply RELEASE_NOTES.txt]: Done.

File: RELEASE_NOTES.txt (hunk 4 above)
Comment (NSAmelchev) [reply RELEASE_NOTES.txt]: Done.

File: RELEASE_NOTES.txt [hunk 7]
```diff
@@ -7,30 +7,27 @@ Apache Ignite In-Memory Distributed Database 2.16.0
...
 * Added logging of snapshot check procedure.
 * Added metrics for operations on secondary indexes.
 * Added node Consistent ID to the index validate command output.
 * Added number of partitions processed by the index worker to the output of the index commands.
+* Added performance statistics for IndexQuery.
 * Added performance statistics info about query properties and processed rows count by query.
 * Added support of SslContextFactory to CommandHandler (instead of GridSslBasicContextFactory).
-* Added support of cache group filter for dumps.
 * Added support of custom SSL factory to the control.sh utility.
 * Added support of custom location for cache dumps.
```
Comment (alex-plekhanov) [inline RELEASE_NOTES.txt:32]: This line related to dumps too.

File: RELEASE_NOTES.txt [hunk 8]
```diff
@@ -1,6 +1,135 @@
 Apache Ignite Release Notes
 ===========================
 
+Apache Ignite In-Memory Distributed Database 2.16.0
+-----------------------------------------------------------
+
+(!) WARNINGS:
+* Operations on atomic caches within transactions are finally forbidden. The system property IGNITE_ALLOW_ATOMIC_OPS_IN_TX has been removed.
+* Removed CacheAtomicityMode#TRANSACTIONAL_SNAPSHOT cache mode.
+* Mixed mode cache groups now restricted, but IGNITE_ALLOW_MIXED_CACHE_GROUPS system option may temporarily allow them.
```
Comment (Sashalig) [inline RELEASE_NOTES.txt:10]: Suggestion: are now restricted

File: RELEASE_NOTES.txt [hunk 9]
```diff
@@ -1,6 +1,135 @@
...
+* The ignite-ml, cassandra modules have been migrated to the Ignite extensions.
```
Comment (Sashalig) [inline RELEASE_NOTES.txt:11]: Suggestion: The ignite-ml and cassandra modules

File: RELEASE_NOTES.txt [hunk 10]
```diff
@@ -1,6 +1,135 @@
...
+
+Ignite:
+* Added ARM64 Docker images.
+* Added CLIENT_CONNECTION_ATTRIBUTES system view.
+* Added JMX commands invoker.
+* Added ability to configure metric exporter filter from Spring XML.
+* Added cache dumps.
+* Added client session outbound message queue size limit.
```
Comment (Sashalig) [inline RELEASE_NOTES.txt:19]: Suggestion: Added size limit of the client session outbound message queue.

File: RELEASE_NOTES.txt [hunk 11]
```diff
@@ -1,6 +1,135 @@
...
+* Fixed Java thin client socket#open timeout to ClientConfiguration#timeout property.
+* Fixed JavaNioAccess handling in JDK 14 and later.
+* Fixed MappedFile.map for JDK 14 and newer.
+* Fixed NPE during defragmentation OOM exception handling.
+* Fixed NPE during transaction recovery.
+* Fixed NPE on snapshot create operation for in-memory cluster.
+* Fixed StoredCacheData removing when a node filter is set.
+* Fixed WAL logging to an archived segment after node restart.
+* Fixed an issue when IgniteLock cannot be acquired after release on JDK 17.
+* Fixed an issue when TTL isn't set for entries inserted with "insert from select" query.
```
Comment (Sashalig) [inline RELEASE_NOTES.txt:49]: Suggestion: is not set

File: RELEASE_NOTES.txt [hunk 12]
```diff
@@ -1,6 +1,135 @@
...
+* Fixed an issue when a transaction will never become finished on timeout when deadlock detection is disabled.
+* Fixed an issue where continuous queries could produce excessive memory usage.
```
Comment (Sashalig) [inline RELEASE_NOTES.txt:51]: Suggestion: Fixed an issue where continuous queries caused excessive memory usage.

File: RELEASE_NOTES.txt [hunk 13]
```diff
@@ -1,6 +1,135 @@
...
+* Fixed assertion error in case of empty SQL query.
+* Fixed assertion error on defragmentation.
+* Fixed broken serialization of communication messages due to incorrect GridCacheQueryRequest class marshalling.
+* Fixed cache objects transformation in Java thin client.
+* Fixed cache scan command page size to meet the limit argument.
```
Comment (Sashalig) [inline RELEASE_NOTES.txt:56]: Suggestion: Fixed size of cache scan command page to meet the limit argument.

File: RELEASE_NOTES.txt [hunk 14]
```diff
@@ -1,6 +1,135 @@
...
+* Fixed connectable nodes list for control.sh.
+* Fixed continuous queries backup ack message sending for expired entries.
+* Fixed control-utility module location for slim and LGPL assemblies.
```
Comment (Sashalig) [inline RELEASE_NOTES.txt:59]: Suggestion: control utility

File: RELEASE_NOTES.txt [hunk 15]
```diff
@@ -1,6 +1,135 @@
...
+* Fixed defragmentation can't be configured to use one thread.
```
Comment (Sashalig) [inline RELEASE_NOTES.txt:60]: Suggestion: Fixed the defragmentation can't be configured to use one thread issue.

File: RELEASE_NOTES.txt [hunk 16]
```diff
@@ -1,6 +1,135 @@
...
+* Fixed erroneous logging of failed index rebuild.
+* Fixed exception message for transaction timeout before data lock acquired.
+* Fixed failure on sqlOnHeapCache=enabled and more than one index used.
+* Fixed histogram metric configuration restore on node restart.
+* Fixed idle_verify hash conflicts for expiring entries.
+* Fixed in-memory server node crashing by TcpIgniteClient.putAllConflict() if cache objects transformation applied.
+* Fixed log message about sending and receiving partition file during snapshot restore.
+* Fixed log message of snapshot create operation with the warning result.
+* Fixed negative duration in the SQL query system view.
+* Fixed node crash due to SecurityContext not being found during discovery message processing.
```
Comment (Sashalig) [inline RELEASE_NOTES.txt:73]: Suggestion: SecurityContext not found during...

File: RELEASE_NOTES.txt [hunk 17]
```diff
@@ -1,6 +1,135 @@
...
+* Added queries timeout support.
+* Added sensitive information hiding when IGNITE_TO_STRING_INCLUDE_SENSITIVE property is set to false.
+* Added support of setting partitions in SqlFieldsQuery.
+* Fixed CREATE TABLE ... WITH affinity_key option validation.
+* Fixed NPE on write query plan to performance statistics.
+* Fixed SQL metrics (added parsing metrics for the Calcite-based engine, fixed execution metrics).
+* Fixed failure on planning when cache size exceeds Integer.MAX_VALUE.
+* Fixed list of reserved keywords.
+* Fixed memory leak in MailboxRegistryImpl#remotes.
+* Fixed performance statistics gathering.
```
Comment (Sashalig) [inline RELEASE_NOTES.txt:115]: Suggestion: performance statistics collection

File: RELEASE_NOTES.txt [hunk 18]
```diff
@@ -1,6 +1,135 @@
...
+* Fixed placeholder values from ModifyNode.
+* Fixed primary key inline size calculation.
+* Fixed query freezes when partitions are set.
```
Comment (Sashalig) [inline RELEASE_NOTES.txt:118]: Suggestion: Fixed the Query freezes when partitions are set issue.

**Review by alex-plekhanov — APPROVED** [review]

**Comment by NSAmelchev** [comment]
@Sashalig, @alex-plekhanov, Thank you for the review!

**Pull request merged**